### PR TITLE
`nil` test parameters should be `nil` in the controller

### PR DESCRIPTION
**Pull request by tenderlove**
This commit fixes a regression where `nil` parameters in the test would
get converted to empty strings in the controller. I added a test to
`4-2-stable` to demonstrate the existing behavior in 3acf0de35533a4d26a1299f63deb2be9875d06c0

This commit will fix the regression, but raises a broader issue about
the behavior of `to_query`.  It seems that our `to_query` implementation
doesn't generate query strings that round trip via Rack's query parser:

```
irb(main):011:0> { :foo => nil }.to_query
=> "foo="
irb(main):012:0> Rack::Utils.default_query_parser.parse_nested_query({ :foo => nil }.to_query)
=> {"foo"=>""}
```

I'm not sure if we *should* fix this, but I guess it's something we need
to know about.

/cc @mistydemeo

### CODE REVIEW

**Comment by rafaelfranca** [comment]
Do you think this would case a different regression now for applications that already changed their tests to match the new behavior?

**Comment by mistydemeo** [comment]
Speaking personally, I adapted my tests by doing `present?` checks instead of `nil?` or truthiness checks, which should apply in both cases.

**Comment by tenderlove** [comment]
@rafaelfranca I don't think so.  To fix this you'd have two options: 1) completely remove the `nil` param from your test, or 2) add `.present?` calls to your controller. I don't think this patch would impact either case.

File: actionpack/lib/action_controller/test_case.rb [hunk 1]
```diff
@@ -90,6 +90,7 @@ def assign_parameters(routes, controller_path, action, parameters, generated_pat
 
       if get?
         if query_string.blank?
+          non_path_parameters.delete_if { |_,v| v.nil? }
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/test_case.rb]: missing space after the comma

**Review by rafaelfranca — APPROVED** [review]

**Comment by matthewd** [comment]
AIUI this is expected behaviour: if your test isn't using something like JSON to encode the parameters, it's impossible to encode nil.. you're simulating a request that no browser could make.

I believe this was a known (and I thought documented) change in behaviour, not a regression.

**Comment by tenderlove** [comment]
@matthewd I'm not sure if it's a known or documented.  I bisected to 394b7be0368cdf992b29c40bf7f501418d630b64 (yes, it's my fault, ugh).  The browser can make such a request, it just omits the parameter.  Presence of the parameter gives you an empty string.  I agree that it doesn't make sense to write a test passing `nil` for a parameter, but a) we don't need to break existing code, b) this only happens with `ActionController::TestCase` which we're actively discouraging, and c) `url_for` will strip `nil` params.  IOW, this test will pass:

```diff
diff --git a/actionpack/test/controller/routing_test.rb b/actionpack/test/controller/routing_test.rb
index ec939e946a..061a7fbb98 100644
--- a/actionpack/test/controller/routing_test.rb
+++ b/actionpack/test/controller/routing_test.rb
@@ -597,6 +597,13 @@ def test_non_controllers_cannot_be_matched
     assert_raise(ActionController::RoutingError) { rs.recognize_path("/not_a/show/10") }
   end
 
+  def test_nil_param_encoding
+    rs.draw do
+      get "post/:id" => "post#show", :as => "post"
+    end
+    assert_equal '/post/1/', url_for(rs, controller: "post", action: "show", id: 1, lol: nil, use_route: "post")
+  end
+
   def test_should_list_options_diff_when_routing_constraints_dont_match
     rs.draw do
       get "post/:id" => "post#show", :constraints => { id: /\d+/ }, :as => "post"
```

**Comment by matthewd** [comment]
If the caller meant to omit the parameter, they wouldn't've supplied it :smile:

My claim is that this can't be a regression, because the previous behaviour was actively broken, in exactly the way that commit intended to fix: it resulted in the key being in `params`, with a value of nil.

While I do feel that key-set-to-empty-string is a closer approximation of key-set-to-nil than key-absent\*, I think the stronger argument at this point is just that that's the way it is: this behaviour is years old, and has been present for the entire 5.x cycle. We can't change the behaviour in the back-branches, and I don't think it's worth the disruption to change it again in a new release.

\* I see the argument for matching `url_for`... but in my mind, the parameter hash approximates a form; a form containing `text_field_tag("foo", nil)` will produce a form submission with `foo=`.

**Comment by al2o3cr** [comment]
@matthewd It's possible to produce this with a query string - `'foo'` parses with `parse_nested_query` as `{ "foo" => nil }` vs `"foo="` which gives `{ "foo" => "" }`.

Stdlib is also inconsistent in this - `URI.encode_www_form` gives the same behavior as above, but `URI.decode_www_form` decodes both `foo` and `foo=` to `[["foo", ""]]`

**Comment by sikachu** [comment]
I feel like this is an expected behavior though, as there's no such thing as `nil` in query parameter value. If a browser sending `?rails=` in the URL, I think the expected result is to get `params` with `{ "rails" => "" }` There's no way the browser could send in any value that would represent nil.

I'd say it's good thing that we investigated and document this, but I'd rather leave this be and consider the current behavior correct going forward. (a.k.a. wontfix)

**Pull request closed without merging**